File: src/0205_isomorphic_strings.py

# LeetCode 205: Isomorphic Strings
# Python version

def _alpha_index(value: str) -> int:
    if value == "a":
        return 0
    if value == "b":
        return 1
    if value == "c":
        return 2
    if value == "d":
        return 3
    if value == "e":
        return 4
    if value == "f":
        return 5
    if value == "g":
        return 6
    if value == "h":
        return 7
    if value == "i":
        return 8
    if value == "j":
        return 9
    if value == "k":
        return 10
    if value == "l":
        return 11
    if value == "m":
        return 12
    if value == "n":
        return 13
    if value == "o":
        return 14
    if value == "p":
        return 15
    if value == "q":
        return 16
    if value == "r":
        return 17
    if value == "s":
        return 18
    if value == "t":
        return 19
    if value == "u":
        return 20
    if value == "v":
        return 21
    if value == "w":
        return 22
    if value == "x":
        return 23
    if value == "y":
        return 24
    if value == "z":
        return 25
    return -1

def isIsomorphic(s: str, t: str) -> bool:
    if len(s) != len(t):
        return False

    map_st = [-1] * 26
    map_ts = [-1] * 26

    for index in range(len(s)):
        left = _alpha_index(s[index])
        right = _alpha_index(t[index])
        if left < 0 or right < 0:
            return False
        if map_st[left] != -1 and map_st[left] != right:
            return False
        if map_ts[right] != -1 and map_ts[right] != left:
            return False
        map_st[left] = right
        map_ts[right] = left

    return True


def main():
    assert isIsomorphic("egg", "add") == True
    assert isIsomorphic("foo", "bar") == False

if __name__ == "__main__":
    main()

File: src/0205_isomorphic_strings.js
var isIsomorphic = function (s, t) {
    if (s.length !== t.length) return false;

    const mapOne = new Map();
    const mapTwo = new Map();

    for (let i = 0; i < s.length; i++) {
        if (mapOne.has(s[i])) {
            if (mapOne.get(s[i]) !== t[i]) return false;
        } else mapOne.set(s[i], t[i]);

        if (mapTwo.has(t[i])) {
            if (mapTwo.get(t[i]) !== s[i]) return false;
        } else mapTwo.set(t[i], s[i]);
    }

    return true;
};

module.exports = { isIsomorphic };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(isIsomorphic("egg", "add"), true);
    assert.deepStrictEqual(isIsomorphic("foo", "bar"), false);
}
